### PR TITLE
feat(manager/nuget): Restore all dependent project files to generate new lock files

### LINKED ISSUE #8010: NuGet lock file not updated for referencing projects
**What Renovate type, platform and version are you using?**

Self hosted v24.2.5

**Describe the bug**

When an update found for a dependency in a project and a lock file exists, Renovate will run `dotnet restore` for that project to update the lock file. It should run `dotnet restore` for projects referencing the updated project as well, since their lock file also needs to be updated.

**Relevant debug logs**

Important part is that it's running the command `dotnet restore src/My.Project/My.Project.csproj`. Later in the PR pipeline `dotnet restore --locked-mode` is failing for a tests project that's referencing `My.Project`.

```
DEBUG: Starting search at index 290 (repository=**repo-name**, packageFile=src/My.Project/My.Project.csproj, branch=renovate/identitymodel-4.x)
       "depName": "IdentityModel"
DEBUG: Found match at index 290 (repository=**repo-name**, packageFile=src/My.Project/My.Project.csproj, branch=renovate/identitymodel-4.x)
       "depName": "IdentityModel"
DEBUG: found NuGet.config (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "nuGetConfigPath": "/tmp/renovate/repos/bitbucket/**repo-name**/nuget.config"
DEBUG: clearing registry URLs (repository=**repo-name**, branch=renovate/identitymodel-4.x)
DEBUG: adding registry URL (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "registryUrl": "https://api.nuget.org/v3/index.json"
DEBUG: adding registry URL (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "registryUrl": "**private registry url**"
DEBUG: Contents updated (repository=**repo-name**, packageFile=src/My.Project/My.Project.csproj, branch=renovate/identitymodel-4.x)
       "depName": "IdentityModel"
DEBUG: nuget.updateArtifacts(src/My.Project/My.Project.csproj) (repository=**repo-name**, branch=renovate/identitymodel-4.x)
DEBUG: found NuGet.config (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "nuGetConfigPath": "/tmp/renovate/repos/bitbucket/**repo-name**/nuget.config"
DEBUG: clearing registry URLs (repository=**repo-name**, branch=renovate/identitymodel-4.x)
DEBUG: adding registry URL (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "registryUrl": "https://api.nuget.org/v3/index.json"
DEBUG: adding registry URL (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "registryUrl": "**private registry url**"
DEBUG: dotnet command (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "cmd": [
         "dotnet nuget add source https://api.nuget.org/v3/index.json --configfile /tmp/renovate/cache/others/nuget/52d61b6cc0e1e555/nuget.config --name nuget.org",
         "dotnet nuget add source **private registry url** --configfile /tmp/renovate/cache/others/nuget/52d61b6cc0e1e555/nuget.config --name private --username pipelines --password **redacted** --store-password-in-clear-text",
         "dotnet restore src/My.Project/My.Project.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/52d61b6cc0e1e555/nuget.config"
       ]
DEBUG: Executing command (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "command": "dotnet nuget add source https://api.nuget.org/v3/index.json --configfile /tmp/renovate/cache/others/nuget/52d61b6cc0e1e555/nuget.config --name nuget.org"
DEBUG: exec completed (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "cmd": "dotnet nuget add source https://api.nuget.org/v3/index.json --configfile /tmp/renovate/cache/others/nuget/52d61b6cc0e1e555/nuget.config --name nuget.org",
       "durationMs": 605,
       "stdout": "Package source with Name: nuget.org added successfully.\n",
       "stderr": ""
DEBUG: Executing command (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "command": "dotnet nuget add source **private registry url** --configfile /tmp/renovate/cache/others/nuget/52d61b6cc0e1e555/nuget.config --name private --username pipelines --password **redacted** --store-password-in-clear-text"
DEBUG: exec completed (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "cmd": "dotnet nuget add source **private registry url** --configfile /tmp/renovate/cache/others/nuget/52d61b6cc0e1e555/nuget.config --name private --username pipelines --password **redacted** --store-password-in-clear-text",
       "durationMs": 578,
       "stdout": "Package source with Name: private added successfully.\n",
       "stderr": ""
DEBUG: Executing command (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "command": "dotnet restore src/My.Project/My.Project.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/52d61b6cc0e1e555/nuget.config"
DEBUG: exec completed (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "cmd": "dotnet restore src/My.Project/My.Project.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/52d61b6cc0e1e555/nuget.config",
       "durationMs": 2734,
       "stdout": "  Determining projects to restore...\n  Restored /tmp/renovate/repos/bitbucket/**repo-name**/src/My.Project/My.Project.csproj (in 1.52 sec).\n",
       "stderr": ""
DEBUG: Returning updated lock file (repository=**repo-name**, branch=renovate/identitymodel-4.x)
DEBUG: Updated 1 package files (repository=**repo-name**, branch=renovate/identitymodel-4.x)
DEBUG: Getting updated lock files (repository=**repo-name**, branch=renovate/identitymodel-4.x)
DEBUG: lock file dirs (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "dirs": {
         "yarnLockDirs": [],
         "npmLockDirs": [],
         "pnpmShrinkwrapDirs": [],
         "lernaDirs": []
       }
DEBUG: Writing package.json files (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "packageFiles": ["package.json"]
DEBUG: Writing any updated package files (repository=**repo-name**, branch=renovate/identitymodel-4.x)
DEBUG: Updated 1 lock files (repository=**repo-name**, branch=renovate/identitymodel-4.x)
       "updatedArtifacts": ["src/My.Project/packages.lock.json"]
DEBUG: 2 file(s) to commit (repository=**repo-name**, branch=renovate/identitymodel-4.x)
DEBUG: Committing files to branch renovate/identitymodel-4.x (repository=**repo-name**, branch=renovate/identitymodel-4.x)
```

**To Reproduce**

Repro repo: https://github.com/lazyboy1/renovate-nuget-lock-file-repro/pull/2

### PR DESCRIPTION
**Pull request by mattvaughan**
## Changes

This change will `dotnet restore` all the project files that are dependent on the project file being operated on. This regenerates lock files needed for multi-project builds to succeed.

## Context

- Closes  #8010
- closes #12314

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/nuget/package-tree.ts
Comment (viceice) [inline lib/manager/nuget/package-tree.ts:24]: We need types for this!

File: lib/manager/nuget/package-tree.ts
Comment (viceice) [inline lib/manager/nuget/package-tree.ts:23]: Use the package reference instead of searching all files.

Ignore references with variables for now.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/manager/nuget/__fixtures__/circular-reference/two/two.csproj
Comment (mattvaughan) [reply lib/manager/nuget/__fixtures__/circular-reference/two/two.csproj:13]: Thought .NET tooling will throw an error, what's to stop someone from crafting this accidentally without running a build, causing the code to go in circles without exiting?

File: lib/manager/nuget/package-tree.ts
Comment (mattvaughan) [reply lib/manager/nuget/package-tree.ts:23]: In .NET, references are defined in the dependent file. So we need to search all files for references to the project file we are operating on to build the reference DAG.

File: lib/manager/nuget/package-tree.ts
Comment (mattvaughan) [reply lib/manager/nuget/package-tree.ts:22]: There were already some references to `localDir` in the NuGet manager code, but this can be refactored :+1:

File: lib/manager/nuget/package-tree.ts
Comment (mattvaughan) [reply lib/manager/nuget/package-tree.ts:24]: I'm no Typescript expert. Does `ReturnType<typeof Graph>` work here?

The types are defined here: https://github.com/datavis-tech/graph-data-structure/blob/master/index.d.ts

**Review by rarkins — APPROVED** [review]

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs deconflicting

File: lib/manager/nuget/artifacts.ts
Comment (viceice) [reply lib/manager/nuget/artifacts.ts:142]: Please don't resolve comments without code change or comments

File: lib/util/fs/index.ts
Comment (viceice) [inline lib/util/fs/index.ts:171]: I would prefer to use `git.getFiles` and minimatch instead.

@rarkins WDYT?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/manager/nuget/artifacts.ts
Comment (mattvaughan) [reply lib/manager/nuget/artifacts.ts:142]: Hmm this must have been in error. I thought I adopted this change. I'll get this removed.

File: lib/util/fs/index.ts
Comment (mattvaughan) [reply lib/util/fs/index.ts:171]: I can't find the `git.getFiles` function. Would you mind sending me a link?

Comment (viceice) [reply lib/util/fs/index.ts:171]: https://github.com/renovatebot/renovate/blob/f5cc1de36f7d6f3ce7d689a9b6675a930d81337c/lib/util/git/index.ts#L476

File: lib/manager/nuget/package-tree.ts
Comment (viceice) [inline lib/manager/nuget/package-tree.ts:69]: This function looks a little bit ugly. 🤔

Comment (mattvaughan) [reply lib/manager/nuget/package-tree.ts:69]: There might be an existing function that does the same thing.

Basically, it takes a relative path from the frame of one directory, and changes it to a relative path from the frame of another.

So a package reference like `../b/b.csproj` in `/tmp/repo/a/a.csproj` would become `b/b.csproj` (relative to `/tmp/repo`).

Let me see if there's already another function that does this, otherwise I'll see what I can do to make it clearer.

Comment (mattvaughan) [reply lib/manager/nuget/package-tree.ts:69]: A better name might be `reframeRealtivePath`

Comment (mattvaughan) [reply lib/manager/nuget/package-tree.ts]: I cleaned this up a bit and renamed the variables to make it more clear.

**Comment by mattvaughan** [comment]
I'll leave the unresolved conversations open for your feedback. I've addressed them, but feel free to push back if you want further changes. If things look good, please mark them resolved.

**Comment by mattvaughan** [comment]
Updated this branch to be compatible with new paths

**Comment by mattvaughan** [comment]
Wanted to follow up again to see if this can get reviewed again! It's been a bit since any feedback was left on the PR and I want to avoid pinging anyone.

**Review by rarkins — DISMISSED** [review]

File: package.json
Comment (viceice) [inline package.json:167]: I think this can be reverted back to dev deps

File: lib/modules/manager/nuget/package-tree.ts
Comment (viceice) [inline lib/modules/manager/nuget/package-tree.ts:80]: Isn't this also on `upath`? So we can use it as drop-in for `path`?

**Comment by mattvaughan** [comment]
Feedback has been addressed. Let me know if further changes are needed :smile:

File: lib/modules/manager/nuget/artifacts.ts
Comment (viceice) [inline lib/modules/manager/nuget/artifacts.ts:183]: remove

File: lib/modules/manager/nuget/artifacts.ts
Comment (viceice) [inline lib/modules/manager/nuget/artifacts.ts:129]:
```suggestion
  logger.trace(
```
not so important

File: lib/modules/manager/nuget/artifacts.ts
Comment (viceice) [inline lib/modules/manager/nuget/artifacts.ts:171]:
```suggestion
        logger.trace(`Lock file ${lockFileName} is unchanged`);
```

File: lib/modules/manager/nuget/package-tree.ts
Comment (viceice) [inline lib/modules/manager/nuget/package-tree.ts:9]: use jsdoc/ tsdoc comment style

File: lib/modules/manager/nuget/package-tree.ts
Comment (viceice) [inline lib/modules/manager/nuget/package-tree.ts:32]: remove

File: lib/modules/manager/nuget/package-tree.ts
Comment (viceice) [inline lib/modules/manager/nuget/package-tree.ts:97]:
```suggestion
  logger.trace({ filteredPackageFiles }, 'Found package files');
```

**Comment by viceice** [comment]
next time please open PR from private github account or add renovate maintainer to your fork, so we can update the branch. Otherwise it will defer merge

**Review by viceice — APPROVED** [review]
LGTM

How can we support this: https://github.com/renovate-reproductions/dotnet-lock-restore-8010

**Comment by viceice** [comment]
Classifying as feature, as it makes big changes 🙃

**Comment by gitfool** [comment]
@viceice looks like an unrelated failure. Retry?

**Comment by viceice** [comment]
> @viceice looks like an unrelated failure. Retry?

sure, you need to update your branch anyways. it must be up to date before merge.

**Comment by gitfool** [comment]
@mattvaughan ^

**Comment by mattvaughan** [comment]
> @mattvaughan ^

Done :smiley:

**Review by viceice — APPROVED** [review]

**Pull request merged**
